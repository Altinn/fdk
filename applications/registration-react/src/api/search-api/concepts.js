import { getConfig } from '../../config';
import { searchApiGet } from './host';

export const searchConcepts = ({ prefLabel, returnFields }) =>
  searchApiGet({
    url: `${getConfig().conceptApi.host}/api/concepts`,
    params: { preflabel: prefLabel, returnFields },
    authorization: getConfig().conceptApi.authorization
  });
