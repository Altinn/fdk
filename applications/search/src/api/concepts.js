import axios from 'axios';
import get from 'lodash/get';
import { getConfig } from '../config';

export const conceptsUrlBase = () =>
  `${getConfig().conceptApi.host}/api/concepts`;

export const conceptsSearch = params =>
  axios
    .get(conceptsUrlBase(), {
      params,
      headers: { authorization: getConfig().conceptApi.authorization }
    })
    .then(r => r.data);

export const getConcept = id =>
  axios
    .get(`${conceptsUrlBase()}/${id}`, {
      headers: { authorization: getConfig().conceptApi.authorization }
    })
    .then(r => r.data)
    .catch(() => null);

export const extractConcepts = searchResponse =>
  get(searchResponse, ['_embedded', 'concepts']);

export const extractAggregations = searchResponse =>
  get(searchResponse, 'aggregations');

export const extractTotal = searchResponse =>
  get(searchResponse, ['page', 'totalElements']);
