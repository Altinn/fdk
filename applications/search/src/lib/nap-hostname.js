import includes from 'lodash/includes';

const napHostnamesWhitelist = {
  hostnames: [
    'nap.ut1.fellesdatakatalog.brreg.no',
    'nap.st1.fellesdatakatalog.brreg.no',
    'nap.it1.fellesdatakatalog.brreg.no',
    'nap.fellesdatakatalog.brreg.no'
  ]
};

export function isNapHostname(hostname) {
  return includes(napHostnamesWhitelist.hostnames, hostname);
}
