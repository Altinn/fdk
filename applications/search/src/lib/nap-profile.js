import includes from 'lodash/includes';

const napHostnamesWhitelist = {
  hostnames: [
    'nap.ut1.fellesdatakatalog.brreg.no',
    'nap.st1.fellesdatakatalog.brreg.no',
    'nap.it1.fellesdatakatalog.brreg.no',
    'nap.fellesdatakatalog.brreg.no'
  ]
};

export function isNapProfile() {
  return (
    includes(napHostnamesWhitelist.hostnames, window.location.hostname) ||
    localStorage.getItem('napProfile')
  );
}
