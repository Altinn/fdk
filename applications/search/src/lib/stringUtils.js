import _ from 'lodash';

export function removeValue(listAsString, value) {
  return _.without(listAsString.split(','), value).join(',') || undefined; // return undefined instead of empty
}

export function addValue(listAsString, value) {
  return (listAsString || '')
    .split(',')
    .concat(value)
    .filter(v => !!v)
    .join(',');
}
