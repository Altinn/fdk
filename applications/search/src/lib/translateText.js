import qs from 'qs';

import localization from './localization';

export function getTranslateText(
  textObj,
  selectedLanguage = localization.getLanguage()
) {
  if (!textObj) {
    return null;
  }
  return (
    textObj[selectedLanguage] ||
    textObj.nb ||
    textObj.no ||
    textObj.nn ||
    textObj.en ||
    null
  );
}

/**
 * Returns language code from url parameter "lang", if exists.
 * @returns {null}
 */
export function getLanguageFromUrl() {
  const queryObj = qs.parse(window.location.search.substr(1));
  if (queryObj && queryObj.lang) {
    return queryObj.lang;
  }
  return null;
}
