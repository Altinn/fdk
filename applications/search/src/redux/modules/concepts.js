import _ from 'lodash';
import qs from 'qs';

import { fetchActions } from '../fetchActions';
import { conceptsSearchUrl } from '../../api/concepts';

export const CONCEPTS_REQUEST = 'CONCEPTS_REQUEST';
export const CONCEPTS_SUCCESS = 'CONCEPTS_SUCCESS';
export const CONCEPTS_FAILURE = 'CONCEPTS_FAILURE';

const generateQueryKey = query => qs.stringify(query, { skipNulls: true });

function shouldFetch(metaState, queryKey) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      ((metaState.lastFetch || 0) < Date.now() - threshold ||
        metaState.queryKey !== queryKey))
  );
}

export function fetchConceptsIfNeededAction(query) {
  const queryKey = generateQueryKey(query);

  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['concepts', 'meta']), queryKey) &&
    dispatch(
      fetchActions(conceptsSearchUrl(query), [
        { type: CONCEPTS_REQUEST, meta: { queryKey } },
        { type: CONCEPTS_SUCCESS, meta: { queryKey } },
        { type: CONCEPTS_FAILURE, meta: { queryKey } }
      ])
    );
}

const initialState = {};

export function conceptReducer(state = initialState, action) {
  switch (action.type) {
    case CONCEPTS_REQUEST:
      return {
        ...state,
        meta: {
          isFetching: true,
          lastFetch: null,
          queryKey: action.meta.queryKey
        }
      };
    case CONCEPTS_SUCCESS:
      return {
        ...state,
        conceptItems: _.get(action.payload, ['_embedded', 'concepts']),
        conceptAggregations: _.get(action.payload, 'aggregations'),
        conceptTotal: _.get(action.payload, ['page', 'totalElements']),
        meta: {
          isFetching: false,
          lastFetch: Date.now(),
          queryKey: action.meta.queryKey
        }
      };
    case CONCEPTS_FAILURE:
      return {
        ...state,
        conceptItems: null,
        conceptAggregations: null,
        conceptTotal: null,
        meta: {
          isFetching: false,
          lastFetch: null, // retry on error
          queryKey: action.meta.queryKey,
          error: action.payload
        }
      };
    default:
      return state;
  }
}
