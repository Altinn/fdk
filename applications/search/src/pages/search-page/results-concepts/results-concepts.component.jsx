import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Button, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import cx from 'classnames';
import _get from 'lodash/get';
import _capitalize from 'lodash/capitalize';
import _ from 'lodash';

import localization from '../../../lib/localization';
import { ConceptsHitItem } from './concepts-hit-item/concepts-hit-item.component';
import { CompareTerms } from './compare-terms/compare-terms.component';
import { CompareTermModal } from './compare-term-modal/compare-term-modal.component';
import { SearchPublishersTree } from '../search-publishers-tree/search-publishers-tree.component';
import { getTranslateText } from '../../../lib/translateText';

export class ResultsConcepts extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      concepts: []
    };
    this.handleAddConcept = this.handleAddConcept.bind(this);
    this.handleDeleteConcept = this.handleDeleteConcept.bind(this);
  }

  handleAddConcept(term) {
    this.setState({
      concepts: [...this.state.concepts, term]
    });
  }

  handleDeleteConcept(uri) {
    const { concepts } = this.state;
    const filteredItems = concepts.filter(item => item.uri !== uri);
    this.setState({
      concepts: filteredItems
    });
  }

  _renderCompareTerms() {
    const { concepts } = this.state;
    const children = items =>
      items.map(item => {
        const { publisher } = item;
        const publisherPrefLabel =
          getTranslateText(_get(publisher, ['prefLabel'])) ||
          _capitalize(_get(publisher, 'name', ''));

        return (
          <CompareTerms
            key={item.uri}
            uri={item.uri}
            prefLabel={item.prefLabel}
            creator={publisherPrefLabel}
            onDeleteTerm={this.handleDeleteConcept}
          />
        );
      });

    const compareButton = (
      <CompareTermModal
        terms={concepts}
        handleDeleteTerm={this.handleDeleteConcept}
      />
    );

    if (concepts && concepts.length > 0) {
      return (
        <div className="mt-4">
          <h3 className="mb-2">{localization.terms.compareTerms}</h3>
          {children(concepts)}
          {compareButton}
        </div>
      );
    }
    return null;
  }

  _renderTerms() {
    const { conceptItems } = this.props;
    if (_.get(conceptItems, ['_embedded', 'concepts'])) {
      return _.get(conceptItems, ['_embedded', 'concepts']).map(item => (
        <ConceptsHitItem
          key={item.id}
          result={item}
          concepts={this.state.concepts}
          onAddConcept={this.handleAddConcept}
          onDeleteConcept={this.handleDeleteConcept}
        />
      ));
    }
    return null;
  }

  _renderFilterModal() {
    const {
      showFilterModal,
      closeFilterModal,
      onFilterPublisherHierarchy,
      searchQuery,
      publisherArray,
      publishers
    } = this.props;
    return (
      <Modal isOpen={showFilterModal} toggle={closeFilterModal}>
        <ModalHeader toggle={closeFilterModal}>Filter</ModalHeader>
        <ModalBody>
          <div className="search-filters">
            <SearchPublishersTree
              title={localization.facet.organisation}
              filter={publisherArray}
              onFilterPublisherHierarchy={onFilterPublisherHierarchy}
              activeFilter={searchQuery.orgPath}
              publishers={publishers}
            />
          </div>
        </ModalBody>
        <ModalFooter>
          <Button
            className="fdk-button-default fdk-button"
            onClick={closeFilterModal}
          >
            Close
          </Button>
        </ModalFooter>
      </Modal>
    );
  }

  render() {
    const {
      conceptItems,
      onClearFilters,
      onPageChange,
      onFilterPublisherHierarchy,
      searchQuery,
      showClearFilterButton,
      hitsPerPage,
      publisherArray,
      publishers
    } = this.props;
    const page = _.get(searchQuery, 'from')
      ? searchQuery.from / hitsPerPage
      : 0;
    const pageCount = Math.ceil(
      (_.get(conceptItems, ['page', 'totalElements'])
        ? conceptItems.page.totalElements
        : 1) / hitsPerPage
    );

    const clearButtonClass = cx(
      'btn',
      'btn-primary',
      'fdk-button',
      'fade-in-500',
      {
        'd-none': !showClearFilterButton
      }
    );

    return (
      <main id="content">
        <section className="row mb-3 fdk-button-row">
          <div className="col-lg-4">
            <button
              className={clearButtonClass}
              onClick={onClearFilters}
              type="button"
            >
              {localization.query.clear}
            </button>
          </div>
        </section>

        <section className="row">
          <aside className="search-filters col-lg-4">
            <div className="d-none d-lg-block">
              <span className="uu-invisible" aria-hidden="false">
                Filtrering tilgang
              </span>
              {_.get(conceptItems, 'aggregations') && (
                <div>
                  {this._renderFilterModal()}
                  <SearchPublishersTree
                    title={localization.facet.organisation}
                    filter={publisherArray}
                    onFilterPublisherHierarchy={onFilterPublisherHierarchy}
                    activeFilter={_.get(searchQuery, 'orgPath')}
                    publishers={publishers}
                  />
                </div>
              )}
            </div>
            {this._renderCompareTerms()}
          </aside>

          <section className="col-lg-8">{this._renderTerms()}</section>

          {_.get(conceptItems, ['page', 'totalElements'], 0) > 50 && (
            <section className="col-lg-8 offset-lg-4 d-flex justify-content-center">
              <span className="uu-invisible" aria-hidden="false">
                Sidepaginering.
              </span>
              <ReactPaginate
                pageCount={pageCount}
                pageRangeDisplayed={2}
                marginPagesDisplayed={1}
                previousLabel={localization.page.prev}
                nextLabel={localization.page.next}
                breakLabel={<span>...</span>}
                breakClassName="break-me"
                containerClassName="pagination"
                onPageChange={onPageChange}
                subContainerClassName="pages pagination"
                activeClassName="active"
                initialPage={page}
                disableInitialCallback
              />
            </section>
          )}
        </section>
      </main>
    );
  }
}

ResultsConcepts.defaultProps = {
  conceptItems: null,
  onClearFilters: null,
  onPageChange: null,
  onFilterPublisherHierarchy: null,
  searchQuery: null,
  hitsPerPage: null,
  showFilterModal: null,
  closeFilterModal: null,
  showClearFilterButton: null,
  publisherArray: null,
  publishers: null
};

ResultsConcepts.propTypes = {
  conceptItems: PropTypes.object,
  onClearFilters: PropTypes.func,
  onPageChange: PropTypes.func,
  onFilterPublisherHierarchy: PropTypes.func,
  searchQuery: PropTypes.object,
  hitsPerPage: PropTypes.number,
  showFilterModal: PropTypes.bool,
  closeFilterModal: PropTypes.func,
  showClearFilterButton: PropTypes.bool,
  publisherArray: PropTypes.array,
  publishers: PropTypes.object
};
