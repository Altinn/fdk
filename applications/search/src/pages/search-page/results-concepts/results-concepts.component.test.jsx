import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import { ResultsConceptsPure } from './results-concepts.component';
import concepts from '../../../../test/fixtures/concepts';
import { normalizeAggregations } from '../../../lib/normalizeAggregations';

test('should render ResultsConcepts correctly with minimum of props', () => {
  const result = shallow(<ResultsConceptsPure />);
  expect(result).toMatchSnapshot();
});

test('should render ResultsConcepts correctly with hits', () => {
  const props = {
    conceptItems: _.get(concepts, ['hits', 'hits']),
    conceptAggregations: _.get(normalizeAggregations(concepts), 'aggregations'),
    conceptTotal: _.get(concepts, ['hits', 'total'])
  };

  const result = shallow(<ResultsConceptsPure {...props} />);
  expect(result).toMatchSnapshot();
});
