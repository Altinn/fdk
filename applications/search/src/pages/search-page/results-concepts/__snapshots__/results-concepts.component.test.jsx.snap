// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ResultsConcepts correctly with minimum of props 1`] = `
<main
  id="content"
>
  <section
    className="row mb-3 fdk-button-row"
  >
    <div
      className="col-6 col-lg-4"
    >
      <button
        className="btn btn-primary fdk-button fade-in-500 d-none"
        onClick={null}
        type="button"
      >
        Clear
      </button>
    </div>
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </section>
  <section
    className="row"
  >
    <aside
      className="search-filters col-lg-4"
    >
      <div
        className="d-none d-lg-block"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Filtrering tilgang
        </span>
      </div>
    </aside>
    <section
      className="col-lg-8"
    />
  </section>
</main>
`;

exports[`should render ResultsConcepts correctly with props 1`] = `
<main
  id="content"
>
  <section
    className="row mb-3 fdk-button-row"
  >
    <div
      className="col-6 col-lg-4"
    >
      <button
        className="btn btn-primary fdk-button fade-in-500 d-none"
        onClick={[MockFunction]}
        type="button"
      >
        Clear
      </button>
    </div>
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </section>
  <section
    className="row"
  >
    <aside
      className="search-filters col-lg-4"
    >
      <div
        className="d-none d-lg-block"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Filtrering tilgang
        </span>
      </div>
    </aside>
    <section
      className="col-lg-8"
    />
  </section>
</main>
`;
